lib/api/src/externals/function.rs

```diff
@@ -67,6 +67,53 @@ pub struct Function {
     pub(crate) exported: ExportFunction,
 }
 
+#[ cfg(feature="async") ]
+fn build_export_function_metadata<Env>(
+    env: Env,
+    import_init_function_ptr: for<'a> fn(
+        &'a mut Env,
+        &'a crate::Instance,
+    ) -> Result<(), crate::HostEnvInitError>,
+    host_env_set_yielder_fn: fn(*mut std::ffi::c_void, *const std::ffi::c_void)
+) -> (*mut std::ffi::c_void, ExportFunctionMetadata)
+where
+    Env: Clone + Sized + 'static + Send + Sync,
+{
+    let import_init_function_ptr = Some(unsafe {
+        std::mem::transmute::<fn(_, _) -> Result<(), _>, fn(_, _) -> Result<(), _>>(
+            import_init_function_ptr,
+        )
+    });
+    let host_env_clone_fn: fn(*mut std::ffi::c_void) -> *mut std::ffi::c_void = |ptr| {
+        let env_ref: &Env = unsafe {
+            ptr.cast::<Env>()
+                .as_ref()
+                .expect("`ptr` to the environment is null when cloning it")
+        };
+        Box::into_raw(Box::new(env_ref.clone())) as _
+    };
+    let host_env_drop_fn: fn(*mut std::ffi::c_void) = |ptr| {
+        unsafe { Box::from_raw(ptr.cast::<Env>()) };
+    };
+    let env = Box::into_raw(Box::new(env)) as _;
+
+    // # Safety
+    // - All these functions work on all threads
+    // - The host env is `Send`.
+    let metadata = unsafe {
+        ExportFunctionMetadata::new(
+            env,
+            import_init_function_ptr,
+            host_env_clone_fn,
+            host_env_set_yielder_fn,
+            host_env_drop_fn,
+        )
+    };
+
+    (env, metadata)
+}
+
+#[ cfg(not(feature="async")) ]
```

Comment:
I'd prefer having the `cfg` logic internal to the original function rather than having 2 copies just because it's so precise, dangerous, and error-prone. The reason this function exists in the first place is to try to avoid duplicating this kind of code.